internal/pkg/skuba/addons/addons.go

```diff
@@ -0,0 +1,288 @@
+		// TODO (ereslibre): this logic can be triggered if some registered
+		//       addons are not supported in all Kubernetes versions. Either:
+		//
+		//   a) When rendering all addons on `skuba cluster init`, we skip those
+		//      that don't apply to the chosen Kubernetes version.
+		//
+		//   b) When running `skuba addon upgrade apply`; in this case (hence the
+		//      TODO), should we trigger a deletion of the addons that are not present
+		//      in the new version but were present on the old Kubernetes version? For
+		//      now, just return that it doesn't have to be applied.
+		return false, nil
+	}
+	if skubaConfiguration.AddonsVersion == nil {
+		return true, nil
+	}
+	currentAddonVersion, found := skubaConfiguration.AddonsVersion[addon.addon]
+	if !found {
+		return true, nil
+	}
+	addonVersion := kubernetes.AddonVersionForClusterVersion(addon.addon, addonConfiguration.ClusterVersion)
+	return !addonVersionLower(addonVersion, currentAddonVersion), nil
+}
+
+func (addon Addon) needsRender(applyBehavior ApplyBehavior) bool {
+	if applyBehavior == AlwaysRender {
+		return true
+	}
+	_, err := os.Stat(addon.manifestPath())
+	return err != nil
+}
+
+func (addon Addon) addonPath() string {
+	return filepath.Join(skubaconstants.AddonsDir(), string(addon.addon))
+}
+
+func (addon Addon) manifestPath() string {
+	return filepath.Join(addon.addonPath(), fmt.Sprintf("%s.yaml", addon.addon))
+}
+
+func (addon Addon) Write(addonConfiguration AddonConfiguration) error {
+	addonManifest, err := addon.Render(addonConfiguration)
+	if err != nil {
+		return errors.Wrapf(err, "unable to render %s addon template", addon.addon)
+	}
+	os.MkdirAll(addon.addonPath(), 0700)
+	if err := ioutil.WriteFile(addon.manifestPath(), []byte(addonManifest), 0600); err != nil {
+		return errors.Wrapf(err, "unable to write %s addon rendered template", addon.addon)
+	}
+	return nil
+}
+
+func (addon Addon) Apply(addonConfiguration AddonConfiguration, skubaConfiguration *skuba.SkubaConfiguration, applyBehavior ApplyBehavior) error {
+	klog.V(1).Infof("applying %q addon", addon.addon)
+	if addon.callbacks != nil {
+		if err := addon.callbacks.BeforeApply(addonConfiguration, skubaConfiguration); err != nil {
+			klog.Errorf("failed on %q addon BeforeApply callback: %v", addon.addon, err)
+			return err
+		}
+	}
+	var renderedManifest string
+	if addon.needsRender(applyBehavior) {
+		var err error
+		renderedManifest, err = addon.Render(addonConfiguration)
+		if err != nil {
+			return err
+		}
+	} else {
+		renderedManifestBytes, err := ioutil.ReadFile(addon.manifestPath())
+		if err != nil {
+			return err
+		}
+		renderedManifest = string(renderedManifestBytes)
+	}
+	cmd := exec.Command("kubectl", "apply", "--kubeconfig", skubaconstants.KubeConfigAdminFile(), "-f", "-")
+	stdin, err := cmd.StdinPipe()
+	if err != nil {
+		return err
+	}
+	if _, err := stdin.Write([]byte(renderedManifest)); err != nil {
+		return err
+	}
+	stdin.Close()
```

Comment:
I'd move this line as a defer before this `if` statement. The reason being: what happens if we couldn't write for whatever reason but the stdin is not really closed? Then with the current code we'd get out with an error but stdin wouldn't be closed.

Reply:
Let me check, because I think we want to close the `stdin` before executing the command (before the `CombinedOutput()` call), and the `defer` would be executed after that.

Reply:
In that case, you need to also add `stdin.Close()` before the `return err` above :smile:

Reply:
Updated to not need a `StdinPipe` but pass the input on `cmd.Stdin` instead :)